lib/msal-core/src/cache/AuthCache.ts

```diff
@@ -0,0 +1,189 @@
+                return key;
+            } else {
+                return addInstanceId ? `${CacheKeys.PREFIX}.${this.clientId}.${key}` : `${CacheKeys.PREFIX}.${key}`;
+            }
+        }
+    }
+
+    // add value to storage
+    setItem(key: string, value: string, enableCookieStorage?: boolean): void {
+        super.setItem(this.generateCacheKey(key, true), value, enableCookieStorage);
+
+        if (this.rollbackEnabled) {
+            super.setItem(this.generateCacheKey(key, false), value, enableCookieStorage);
+        }
+    }
+
+    // get one item by key from storage
+    getItem(key: string, enableCookieStorage?: boolean): string {
+        return super.getItem(this.generateCacheKey(key, true), enableCookieStorage);
+    }
+
+    // remove value from storage
+    removeItem(key: string): void {
+        super.removeItem(this.generateCacheKey(key, true));
+        if (this.rollbackEnabled) {
+            super.removeItem(this.generateCacheKey(key, false));
+        }
+    }
+
+    resetCacheItems(): void {
+        const storage = window[this.cacheLocation];
+        let key: string;
+        this.removeAcquireTokenEntries();
+        for (key in storage) {
+            if (storage.hasOwnProperty(key)) {
+                // Check if key contains msal prefix
+                if (key.indexOf(CacheKeys.PREFIX) !== -1) {
+                    // For now, we are clearing all cache items created by MSAL.js
+                    super.removeItem(key);
+                    // TODO: Clear cache based on client id (clarify use cases where this is needed)
+                }
+            }
+        }
+    }
+
+    setItemCookie(cName: string, cValue: string, expires?: number): void {
+        super.setItemCookie(this.generateCacheKey(cName, true), cValue, expires);
+        if (this.rollbackEnabled) {
+            super.setItemCookie(this.generateCacheKey(cName, false), cValue, expires);
+        }
+    }
+
+    getItemCookie(cName: string): string {
+        return super.getItemCookie(this.generateCacheKey(cName, true));
+    }
+
+    getAllAccessTokens(clientId: string, homeAccountIdentifier: string): Array<AccessTokenCacheItem> {
+        const results: Array<AccessTokenCacheItem> = [];
+        let accessTokenCacheItem: AccessTokenCacheItem;
+        const storage = window[this.cacheLocation];
+        if (storage) {
+            let key: string;
+            for (key in storage) {
+                if (storage.hasOwnProperty(key)) {
+                    if (key.match(clientId) && key.match(homeAccountIdentifier)) {
+                        const value = this.getItem(key);
+                        if (value) {
+                            accessTokenCacheItem = new AccessTokenCacheItem(JSON.parse(key), JSON.parse(value));
+                            results.push(accessTokenCacheItem);
+                        }
+                    }
+                }
+            }
+        }
+
+        return results;
```

Comment:
This could be done with `reduce`,

```js

const results = Object.keys(window[this.cacheLocation]).reduce((tokens, key) => {
    if (storage.hasOwnProperty(key) && key.match(clientId) && key.match(homeAccountIdentifier)) {
        const value = this.getItem(key);
        if (value) {
           const newAccessTokenCacheItem = new AccessTokenCacheItem(JSON.parse(key), JSON.parse(value));
           return tokens.concat([ newAccessTokenCacheItem ]);
        }
     }

     return tokens;
}, []);

return results;
```

Reply:
+1